### ISSUE TITLE
[BUG] Failed to run pyscenic grn on 28k cells

### ISSUE THREAD

**Issue by YiqunCao** [aertslab/pySCENIC#238]
Hi, 

- pySCENIC version: 0.10.4
- Installation method: Pip
- Run environment: Jupyter notebook
- OS: Ubuntu

When I try to run grn on 28k cells, the following error occurs. But when I subset the dataset to 3k cells and re-run the same thing, no such error anymore. What could be the problem? Thanks! 

```python
!pyscenic grn {f_loom_path_scenic} {f_tfs} -o adj.csv --num_workers 5
```

```pytb
'infer_data failed for target ZNF252P-AS1' Retry (1/10). Failure caused by ValueError("Regression for target gene ZNF252P-AS1 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TENT5A' Retry (1/10). Failure caused by ValueError("Regression for target gene TENT5A failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TTPAL' Retry (1/10). Failure caused by ValueError("Regression for target gene TTPAL failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TMEM245' Retry (1/10). Failure caused by ValueError("Regression for target gene TMEM245 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZDHHC2' Retry (1/10). Failure caused by ValueError("Regression for target gene ZDHHC2 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZNF252P-AS1' Retry (2/10). Failure caused by ValueError("Regression for target gene ZNF252P-AS1 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TENT5A' Retry (2/10). Failure caused by ValueError("Regression for target gene TENT5A failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TMEM245' Retry (2/10). Failure caused by ValueError("Regression for target gene TMEM245 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TTPAL' Retry (2/10). Failure caused by ValueError("Regression for target gene TTPAL failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZDHHC2' Retry (2/10). Failure caused by ValueError("Regression for target gene ZDHHC2 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZNF252P-AS1' Retry (3/10). Failure caused by ValueError("Regression for target gene ZNF252P-AS1 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TENT5A' Retry (3/10). Failure caused by ValueError("Regression for target gene TENT5A failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TMEM245' Retry (3/10). Failure caused by ValueError("Regression for target gene TMEM245 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZDHHC2' Retry (3/10). Failure caused by ValueError("Regression for target gene ZDHHC2 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TTPAL' Retry (3/10). Failure caused by ValueError("Regression for target gene TTPAL failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZNF252P-AS1' Retry (4/10). Failure caused by ValueError("Regression for target gene ZNF252P-AS1 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZDHHC2' Retry (4/10). Failure caused by ValueError("Regression for target gene ZDHHC2 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TENT5A' Retry (4/10). Failure caused by ValueError("Regression for target gene TENT5A failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TMEM245' Retry (4/10). Failure caused by ValueError("Regression for target gene TMEM245 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZNF252P-AS1' Retry (5/10). Failure caused by ValueError("Regression for target gene ZNF252P-AS1 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TTPAL' Retry (4/10). Failure caused by ValueError("Regression for target gene TTPAL failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZDHHC2' Retry (5/10). Failure caused by ValueError("Regression for target gene ZDHHC2 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TMEM245' Retry (5/10). Failure caused by ValueError("Regression for target gene TMEM245 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TENT5A' Retry (5/10). Failure caused by ValueError("Regression for target gene TENT5A failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZNF252P-AS1' Retry (6/10). Failure caused by ValueError("Regression for target gene ZNF252P-AS1 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TTPAL' Retry (5/10). Failure caused by ValueError("Regression for target gene TTPAL failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZDHHC2' Retry (6/10). Failure caused by ValueError("Regression for target gene ZDHHC2 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZNF252P-AS1' Retry (7/10). Failure caused by ValueError("Regression for target gene ZNF252P-AS1 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TMEM245' Retry (6/10). Failure caused by ValueError("Regression for target gene TMEM245 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TENT5A' Retry (6/10). Failure caused by ValueError("Regression for target gene TENT5A failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TTPAL' Retry (6/10). Failure caused by ValueError("Regression for target gene TTPAL failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZDHHC2' Retry (7/10). Failure caused by ValueError("Regression for target gene ZDHHC2 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZNF252P-AS1' Retry (8/10). Failure caused by ValueError("Regression for target gene ZNF252P-AS1 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TMEM245' Retry (7/10). Failure caused by ValueError("Regression for target gene TMEM245 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TENT5A' Retry (7/10). Failure caused by ValueError("Regression for target gene TENT5A failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TTPAL' Retry (7/10). Failure caused by ValueError("Regression for target gene TTPAL failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZNF252P-AS1' Retry (9/10). Failure caused by ValueError("Regression for target gene ZNF252P-AS1 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZDHHC2' Retry (8/10). Failure caused by ValueError("Regression for target gene ZDHHC2 failed. Cause ValueError('buffer source array is read-only').").

'infer_data failed for target TMEM245' Retry (8/10). Failure caused by ValueError("Regression for target gene TMEM245 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TENT5A' Retry (8/10). Failure caused by ValueError("Regression for target gene TENT5A failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZNF252P-AS1' Retry (10/10). Failure caused by ValueError("Regression for target gene ZNF252P-AS1 failed. Cause ValueError('buffer source array is read-only').").
distributed.worker - WARNING -  Compute Failed
Function:  execute_task
args:      ((<function check_meta at 0x7f42aa13ad08>, (<function infer_partial_network at 0x7f42822a4840>, 'GBM', (<class 'dict'>, [['learning_rate', 0.01], ['n_estimators', 5000], ['max_features', 0.1], ['subsample', 0.9]]), array([[0.       , 0.       , 0.       , ..., 0.       , 0.       ,
        0.       ],
       [0.       , 0.       , 0.       , ..., 0.       , 0.       ,
        0.       ],
       [0.       , 0.       , 0.       , ..., 0.       , 0.       ,
        0.       ],
       ...,
       [0.       , 0.4185284, 0.       , ..., 0.       , 0.       ,
        0.       ],
       [0.       , 0.       , 0.       , ..., 0.       , 0.       ,
        0.       ],
       [0.       , 0.       , 0.       , ..., 0.       , 0.       ,
        0.       ]], dtype=float32), ['A1CF', 'ABCF2', 'ABL1', 'ACAA1', 'ACO1', 'ADARB1', 'ADNP', 'ADNP2', 'AEBP2', 'AFF4', 'AGGF1', 'AGMAT', 'AHCTF1', 'AHR', 'AHRR', 'AKR1A1', 'ALX1', 'ANXA1', 'ANXA11', 'APEX1', 'APEX2', 'AR', 'ARFGAP1', 'ARG2', 'ARID3A', 'ARID3B'
kwargs:    {}
Exception: ValueError('Metadata mismatch found in `from_delayed`.\n\nExpected partition of type `DataFrame` but got `NoneType`')

'infer_data failed for target TTPAL' Retry (8/10). Failure caused by ValueError("Regression for target gene TTPAL failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZDHHC2' Retry (9/10). Failure caused by ValueError("Regression for target gene ZDHHC2 failed. Cause ValueError('buffer source array is read-only').").
not shutting down client, client was created externally
finished
'infer_data failed for target TMEM245' Retry (9/10). Failure caused by ValueError("Regression for target gene TMEM245 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TENT5A' Retry (9/10). Failure caused by ValueError("Regression for target gene TENT5A failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TRIM5' Retry (1/10). Failure caused by ValueError("Regression for target gene TRIM5 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target ZDHHC2' Retry (10/10). Failure caused by ValueError("Regression for target gene ZDHHC2 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TTPAL' Retry (9/10). Failure caused by ValueError("Regression for target gene TTPAL failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TRIM5' Retry (2/10). Failure caused by ValueError("Regression for target gene TRIM5 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TMEM245' Retry (10/10). Failure caused by ValueError("Regression for target gene TMEM245 failed. Cause ValueError('buffer source array is read-only').").
'infer_data failed for target TENT5A' Retry (10/10). Failure caused by ValueError("Regression for target gene TENT5A failed. Cause ValueError('buffer source array is read-only').").
Traceback (most recent call last):
  File "/home/yiqun/anaconda3/bin/pyscenic", line 10, in <module>
    sys.exit(main())
  File "/home/yiqun/anaconda3/lib/python3.7/site-packages/pyscenic/cli/pyscenic.py", line 470, in main
    args.func(args)
  File "/home/yiqun/anaconda3/lib/python3.7/site-packages/pyscenic/cli/pyscenic.py", line 74, in find_adjacencies_command
    network = method(expression_data=ex_mtx, tf_names=tf_names, verbose=True, client_or_address=client, seed=args.seed)
  File "/home/yiqun/anaconda3/lib/python3.7/site-packages/arboreto/algo.py", line 41, in grnboost2
    early_stop_window_length=early_stop_window_length, limit=limit, seed=seed, verbose=verbose)
  File "/home/yiqun/anaconda3/lib/python3.7/site-packages/arboreto/algo.py", line 135, in diy
    .compute(graph, sync=True) \
  File "/home/yiqun/anaconda3/lib/python3.7/site-packages/distributed/client.py", line 2758, in compute
    result = self.gather(futures)
  File "/home/yiqun/anaconda3/lib/python3.7/site-packages/distributed/client.py", line 1822, in gather
    asynchronous=asynchronous,
  File "/home/yiqun/anaconda3/lib/python3.7/site-packages/distributed/client.py", line 753, in sync
    return sync(self.loop, func, *args, **kwargs)
  File "/home/yiqun/anaconda3/lib/python3.7/site-packages/distributed/utils.py", line 331, in sync
    six.reraise(*error[0])
  File "/home/yiqun/anaconda3/lib/python3.7/site-packages/six.py", line 703, in reraise
    raise value
  File "/home/yiqun/anaconda3/lib/python3.7/site-packages/distributed/utils.py", line 316, in f
    result[0] = yield future
  File "/home/yiqun/anaconda3/lib/python3.7/site-packages/tornado/gen.py", line 735, in run
    value = future.result()
  File "/home/yiqun/anaconda3/lib/python3.7/site-packages/tornado/gen.py", line 742, in run
    yielded = self.gen.throw(*exc_info)  # type: ignore
  File "/home/yiqun/anaconda3/lib/python3.7/site-packages/distributed/client.py", line 1653, in _gather
    six.reraise(type(exception), exception, traceback)
  File "/home/yiqun/anaconda3/lib/python3.7/site-packages/six.py", line 702, in reraise
    raise value.with_traceback(tb)
  File "/home/yiqun/anaconda3/lib/python3.7/site-packages/dask/dataframe/utils.py", line 521, in check_meta
    errmsg))
ValueError: Metadata mismatch found in `from_delayed`.

Expected partition of type `DataFrame` but got `NoneType`
```

**Comment by JingleW** [comment]
I met the same problem. The pipeline worked well for the pbmc data, but not for my own data, which is much bigger.

**Comment by cflerin** [comment]
Hi @YiqunCao ,

Did you filter your expression matrix? Do you have any genes or cells with completely empty expression values?

**Comment by JingleW** [comment]
Hi @cflerin , thanks for the great tool pySCENIC. 
    For my own data, I processed the 10X data using Seurat. Cells with few or high gene count and high mitochondrial gene expression were filtered out. And the count matrix without normalization exported from Seurat were used for pySCENIC. There are some genes or cells with completely empty expression values. And pySCENIC reported the same error as @YiqunCao described. I have two questions:
1. Should the count matrix or the normalized data matrix by Seurat be used for pySCENIC?
2. Must genes with low overall expression be removed? As shown in the Nature Protocols, genes expressed in fewer than three cells are removed. 

Thank you very much.

**Comment by JingleW** [comment]
> Hi @cflerin , thanks for the great tool pySCENIC.
> For my own data, I processed the 10X data using Seurat. Cells with few or high gene count and high mitochondrial gene expression were filtered out. And the count matrix without normalization exported from Seurat were used for pySCENIC. There are some genes or cells with completely empty expression values. And pySCENIC reported the same error as @YiqunCao described. I have two questions:
> 
>     1. Should the count matrix or the normalized data matrix by Seurat be used for pySCENIC?
> 
>     2. Must genes with low overall expression be removed? As shown in the Nature Protocols, genes expressed in fewer than three cells are removed.
> 
> 
> Thank you very much.

My original count matrix is 22244 cells and 33539 genes. I tried several settings to filter out genes in few cells. 
1. Filtering genes expressed in fewer than 3 cells, then the matrix is about 20000 genes, pySCENIC still reported the error. 
2. Filtering genes expressed in less than 0.5% cells, then the matrix is about 12000 genes, pySCENIC still reported the error. 
3. Filtering genes expressed in less than 1% cells, then the matrix is about 11000 genes, pySCENIC worked. 

According to the protocol of SCENIC (not pySCENIC), genes expressed in less than 1% cells were filtered out. I wonder which is the optimal setting for filtering. The same setting, for example 1%, for all the dataset? Or it should be adjusted for different dataset?
Thank you very much.

**Comment by cflerin** [comment]
Hi @YiqunCao , 

I believe the original metadata mismatch issue should be solved in pySCENIC 0.11.0, so please try again with that version. I think my original question on filtering was on the wrong track.

For the other filtering questions, yes definitely filter out the completely empty genes/cells, following the protocol recommendations. There is probably not much difference between the 1% and 3-cell cutoffs in practice, and you will get very little information from these genes anyway. The GRN should run fine with the less stringent threshold, but will of course take longer.

**Comment by Emma1997WTT** [comment]
> 

Thanks!I had the same error,and I switch  pySCENIC 0.10.0 to  pySCENIC 0.11.0,then it worked!
